### DISCUSSION TITLE
Doubt about attention coefficient of gat

### DISCUSSION THREAD

**Discussion by dhorka** [pyg-team/pytorch_geometric#2918] [Q&A] [answered]
Hi,

After checking the current implementation of GAT I have not been able to find the concatenation of [ Whi || Whj] described in the paper  The original formulation says that: 
![Screenshot_20210729_171846](https://user-images.githubusercontent.com/39189267/127518851-bad0087f-4639-4294-a77c-f4a82d6188b1.png)
What I saw is instead of doing a concatenation, the code is doing a summation in the message function `alpha = alpha_j if alpha_i is None else alpha_j + alpha_i`. Is there something that I am missing? I am asking that because I would like to change this concatenation for a dot product, but I was not able to find this concatenation step.

Thanks,

**Comment by rusty1s** [comment, marked as answer]
We concatenate implicitly, leading to a smaller memory footprint. In particular, we hold two versions of the `a` parameter vector, one for `W@h_i` (named `att_r`) and one for `W@h_j` (named `att_l`). We then multiply the source and destination node features with these parameters and sum the resulting parts together. This is equivalent to first concatenating source and destination node features, and multiply with a single attention parameter vector afterwards. Hope this is understandable :)

**Reply by dhorka** [reply]
Hmmm, I am not entirely seeing how adding the results of applying both `att_r` independently is equivalent to the concatenation. Maybe my confusion comes from something that I do not understand correctly in the paper. What I understood is that `a` has an output of `2*F`. If this `2*F` is the size of the parameter vector instead of the output of applying, then I understand what you are saying. Can you confirm that `2*F` is the size of the parameter vector instead of the output?  

On the other hand, If I want to change this concatenation for a dot product. I should use one `att` vector that should be applied to the resulting of doing the dot product between `W@h_i` `W@h_j`. Isn't it?

**Reply by rusty1s** [reply]
`a` has shape `2*F`, which gets multiplied with a `2*F` vector to obtain a scalar output. For applying dot product attention, the [`SuperGATConv` already implements this](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/conv/supergat_conv.py#L253-L272).

**Reply by dhorka** [reply]
Hi,

after checking the code that you have pointed I have not been able to find where the dot product is implemented. What I am trying to do is modify the function `a(concat(Wx_i, Wx_j))` by `dot(Wx_i, Wx_j)`. This modification is in the code that you have sent me? Do you think that you can give me some hint to how to implement this efficiently with the current GAT implementation? 

Thanks!

**Reply by rusty1s** [reply]
Yes, this is in the code, i.e., line 265-268, in which attention coefficients are implemented via `(x_i * x_j).sum(dim=-1)`. You can do the same in `GATConv`, where you simply replace the `alpha` computation in `message` via `alpha = (x_i * x_j).sum(dim=-1) / math.sqrt(self.out_channels)`.

**Reply by dhorka** [reply]
I didn't realize that was the dot product. Sorry, but there is something that I do not understand why are you dividing the product for the sqrt of the number of channels?

**Reply by rusty1s** [reply]
This actually refers to the scaled dot-product attention, as introduced in the ["Attention is All You Need"](https://arxiv.org/abs/1706.03762) paper.

**Reply by dhorka** [reply]
Thanks! All my doubts have been clarified!
